gt/main.c

```diff
@@ -870,14 +794,344 @@ print_unsent_policy(struct ggu_policy *policy)
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_udp->dgram_cksum =
+			rte_ipv6_phdr_cksum(notify_ipv6,
+				ggu_pkt->buf->ol_flags);
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u sending notification packet to Gatekeeper server with unknown IP version %hu\n",
+			rte_lcore_id(), ggu_pkt->ipaddr.proto);
+	}
+
+	notify_udp->dgram_len =
+		rte_cpu_to_be_16((uint16_t)(ggu_pkt->buf->data_len -
+			ggu_pkt->buf->l2_len - ggu_pkt->buf->l3_len));
+}
+
+static void
+send_notify_pkt(struct gt_config *gt_conf, struct gt_instance *instance,
+	struct ggu_notify_pkt *ggu_pkt)
+{
+	prep_notify_pkt(ggu_pkt);
+
+	if (rte_eth_tx_burst(gt_conf->net->front.id,
+			instance->tx_queue, &ggu_pkt->buf, 1) != 1) {
+		print_unsent_policies(ggu_pkt);
+		rte_pktmbuf_free(ggu_pkt->buf);
+	}
+
+	ggu_pkt->buf = NULL;
+	instance->num_ggu_pkts--;
+}
+
+/* Send all saved policy decision notification packets being buffered. */
+static void
+flush_notify_pkts(struct gt_config *gt_conf, struct gt_instance *instance)
+{
+	unsigned int max_pkts = gt_conf->max_ggu_notify_pkts;
+	struct rte_mbuf *bufs[max_pkts];
+	int num_to_send = 0;
+	int num_sent;
+	unsigned int i;
+
+	if (instance->num_ggu_pkts == 0)
+		return;
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+
+		prep_notify_pkt(&instance->ggu_pkts[i]);
+		bufs[num_to_send++] = instance->ggu_pkts[i].buf;
+	}
+
+	num_sent = rte_eth_tx_burst(gt_conf->net->front.id,
+		instance->tx_queue, bufs, num_to_send);
+	if (unlikely(num_sent < num_to_send)) {
+		/*
+		 * Need to exclude the first @num_sent valid entries
+		 * found in instance->ggu_pkts from being printed.
+		 */
+		int sent_count = 0;
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[i].buf == NULL)
+				continue;
+
+			if (sent_count < num_sent) {
+				sent_count++;
+				continue;
+			}
+
+			print_unsent_policies(&instance->ggu_pkts[i]);
+			rte_pktmbuf_free(instance->ggu_pkts[i].buf);
+		}
+	}
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+
+		instance->ggu_pkts[i].buf = NULL;
+		instance->num_ggu_pkts--;
+	}
+
+	RTE_VERIFY(instance->num_ggu_pkts == 0);
+}
+
+/*
+ * Start building a new notification packet for the Gatekeeper
+ * server indicated by @pkt_info.
+ *
+ * If there's no more room for a notification packet, then
+ * send a random one to make room.
+ */
+static struct ggu_notify_pkt *
+add_notify_pkt(struct gt_config *gt_conf, struct gt_instance *instance,
+	struct gt_packet_headers *pkt_info)
+{
+	unsigned int max_pkts = gt_conf->max_ggu_notify_pkts;
+	struct ggu_notify_pkt *ggu_pkt = NULL;
+	unsigned int lcore_id = rte_lcore_id();
+	unsigned int i;
+
+	/* Find an available packet, sending a packet if necessary. */
+	if (instance->num_ggu_pkts == max_pkts) {
+		int idx = rte_rand() % max_pkts;
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[idx].buf != NULL)
```

Comment:
Given that `instance->num_ggu_pkts == max_pkts`, `instance->ggu_pkts[idx].buf` must be not `NULL` without the `for` loop. Otherwise `instance->num_ggu_pkts` is bogus. That is, we don't need the `for` loop here.